src/persist/src/storage.rs

```diff
@@ -622,6 +666,94 @@ pub mod tests {
         Ok(())
     }
 
+    pub async fn consensus_impl_test<C: Consensus, F: FnMut() -> Result<C, Error>>(
+        mut new_fn: F,
+    ) -> Result<(), Error> {
+        let mut consensus = new_fn()?;
+
+        // Enforce that this entire test completes within 10 minutes.
+        let deadline = Instant::now() + Duration::from_secs(600);
+
+        // Starting value of consensus data is None.
+        assert_eq!(consensus.head(deadline).await, Ok(None));
+
+        let new_state = VersionedData {
+            seqno: SeqNo(5),
+            data: "abc".as_bytes().to_vec(),
+        };
+
+        // Incorrectly setting the data with a non-None expected should fail.
+        assert_eq!(
+            consensus
+                .compare_and_set(deadline, Some(SeqNo(0)), new_state.clone())
+                .await,
+            Ok(Err(None))
+        );
+
+        // Correctly updating the state with the correct expected value should succeed.
+        assert_eq!(
+            consensus
+                .compare_and_set(deadline, None, new_state.clone())
+                .await,
+            Ok(Ok(()))
+        );
+
+        // We can observe the a recent value on successful update.
+        assert_eq!(consensus.head(deadline).await, Ok(Some(new_state.clone())));
+
+        let new_state2 = VersionedData {
+            seqno: SeqNo(10),
+            data: "def".as_bytes().to_vec(),
+        };
+
+        // Trying to update without the correct expected seqno fails, (even if expected > current)
+        assert_eq!(
+            consensus
+                .compare_and_set(deadline, Some(SeqNo(7)), new_state2.clone())
+                .await,
+            Ok(Err(Some(new_state.clone())))
+        );
+
+        // Trying to update without the correct expected seqno fails, (even if expected < current)
+        assert_eq!(
+            consensus
+                .compare_and_set(deadline, Some(SeqNo(3)), new_state2.clone())
+                .await,
+            Ok(Err(Some(new_state.clone())))
+        );
+
+        // Can correctly update to a new state if we provide the right expected seqno
+        assert_eq!(
+            consensus
+                .compare_and_set(deadline, Some(new_state.seqno.clone()), new_state2.clone(),)
+                .await,
+            Ok(Ok(()))
+        );
+
+        // We can observe the a recent value on successful update.
+        assert_eq!(consensus.head(deadline).await, Ok(Some(new_state2.clone())));
+
+        // Trying to set the data to a sequence number == current fails even if
+        // expected is correct.
+        assert_eq!(
+            consensus
+                .compare_and_set(deadline, Some(new_state2.seqno.clone()), new_state2.clone(),)
```

Comment:
`new_state2` is the actual current state, but you're using it for the new value here. this makes the assertions below a bit unobvious. probably worth introducing a new_state3 or something for the `new` values. or just move these up to before you successfully set it to new_state2 (i.e. while `head` is still `new_state`)

Reply:
done.